// react-static/lib/static/components/Redirect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _reactHelmet = _interopRequireDefault(require("react-helmet"));

//
var Redirect =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(Redirect, _React$Component);

  function Redirect() {
    (0, _classCallCheck2["default"])(this, Redirect);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Redirect).apply(this, arguments));
  }

  (0, _createClass2["default"])(Redirect, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          to = _this$props.to,
          fromPath = _this$props.fromPath;
      var resolvedTo = (0, _typeof2["default"])(to) === 'object' ? to.pathname : to;

      if (!resolvedTo.includes('//')) {
        resolvedTo = "".concat(process.env.REACT_STATIC_PUBLIC_PATH).concat(resolvedTo === '/' ? '' : resolvedTo);
      }

      return _react["default"].createElement(_reactHelmet["default"], null, fromPath && _react["default"].createElement("title", null, "".concat(process.env.REACT_STATIC_PUBLIC_PATH).concat(fromPath === '/' ? '' : fromPath)), _react["default"].createElement("link", {
        rel: "canonical",
        href: resolvedTo
      }), _react["default"].createElement("meta", {
        name: "robots",
        content: "noindex"
      }), _react["default"].createElement("meta", {
        httpEquiv: "content-type",
        content: "text/html; charset=utf-8"
      }), _react["default"].createElement("meta", {
        httpEquiv: "refresh",
        content: "0; url=".concat(resolvedTo)
      }));
    }
  }]);
  return Redirect;
}(_react["default"].Component);

exports["default"] = Redirect;
