// react-static/lib/static/components/RootComponents.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Body = exports.Head = exports.Html = exports.DefaultDocument = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var DefaultDocument =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(DefaultDocument, _Component);

  function DefaultDocument() {
    (0, _classCallCheck2["default"])(this, DefaultDocument);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DefaultDocument).apply(this, arguments));
  }

  (0, _createClass2["default"])(DefaultDocument, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          Html = _this$props.Html,
          Head = _this$props.Head,
          Body = _this$props.Body,
          children = _this$props.children;
      return _react["default"].createElement(Html, null, _react["default"].createElement(Head, null, _react["default"].createElement("meta", {
        charSet: "UTF-8"
      }), _react["default"].createElement("meta", {
        name: "viewport",
        content: "width=device-width, initial-scale=1, maximum-scale=5, shrink-to-fit=no"
      })), _react["default"].createElement(Body, null, children));
    }
  }]);
  return DefaultDocument;
}(_react.Component);

exports.DefaultDocument = DefaultDocument;

var Html = function Html(_ref) {
  var children = _ref.children,
      rest = (0, _objectWithoutProperties2["default"])(_ref, ["children"]);
  return _react["default"].createElement("html", (0, _extends2["default"])({
    lang: "en-US"
  }, rest), children);
};

exports.Html = Html;

var Head = function Head(_ref2) {
  var children = _ref2.children,
      rest = (0, _objectWithoutProperties2["default"])(_ref2, ["children"]);
  return _react["default"].createElement("head", rest, children);
};

exports.Head = Head;

var Body = function Body(_ref3) {
  var children = _ref3.children,
      rest = (0, _objectWithoutProperties2["default"])(_ref3, ["children"]);
  return _react["default"].createElement("body", rest, children);
};

exports.Body = Body;
