// react-static/lib/utils/createIndexPlaceholder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createIndexPlaceholder;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _react = _interopRequireDefault(require("react"));

var _server = require("react-dom/server");

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _RootComponents = require("../static/components/RootComponents");

function createIndexPlaceholder(_x) {
  return _createIndexPlaceholder.apply(this, arguments);
}

function _createIndexPlaceholder() {
  _createIndexPlaceholder = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee(state) {
    var _state$config, Document, paths, Component, DocumentHtml, html;

    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _state$config = state.config, Document = _state$config.Document, paths = _state$config.paths; // Render the base document component to string with siteprops

            Component = Document || _RootComponents.DefaultDocument;
            DocumentHtml = (0, _server.renderToString)(_react["default"].createElement(Component, {
              Html: _RootComponents.Html,
              Head: _RootComponents.Head,
              Body: _RootComponents.Body,
              state: state
            }, _react["default"].createElement("div", {
              id: "root"
            })));
            html = "<!DOCTYPE html>".concat(DocumentHtml); // Write the Document to index.html

            _context.next = 6;
            return _fsExtra["default"].outputFile(paths.HTML_TEMPLATE, html);

          case 6:
            return _context.abrupt("return", state);

          case 7:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _createIndexPlaceholder.apply(this, arguments);
}
