// react-static/lib/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ChalkColor: true,
  findAvailablePort: true,
  time: true,
  timeEnd: true,
  debounce: true,
  escapeRegExp: true,
  progress: true
};
exports.time = time;
exports.timeEnd = timeEnd;
exports.debounce = debounce;
exports.escapeRegExp = escapeRegExp;
Object.defineProperty(exports, "progress", {
  enumerable: true,
  get: function get() {
    return _progress["default"];
  }
});
exports.findAvailablePort = exports.ChalkColor = void 0;

var _portfinder = _interopRequireDefault(require("portfinder"));

var _perf_hooks = require("perf_hooks");

var _utils = require("../browser/utils");

Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _utils[key];
    }
  });
});

var _progress = _interopRequireDefault(require("./progress"));

//
// Export all shared utils
var ChalkColor = {
  yarn: '#2c8ebb',
  npm: '#cb3837'
};
exports.ChalkColor = ChalkColor;

var findAvailablePort = function findAvailablePort(start) {
  return _portfinder["default"].getPortPromise({
    port: start,
    stopPort: start + 1000
  });
};

exports.findAvailablePort = findAvailablePort;
var times = {};

function time(message) {
  times[message] = _perf_hooks.performance.now() / 1000;
}

function timeEnd(message) {
  if (times[message]) {
    var seconds = (_perf_hooks.performance.now() / 1000 - times[message]) * 10;
    times[message] = null;

    if (seconds < 0.1) {
      console.log("".concat(message));
      return;
    }

    if (seconds < 1) {
      seconds = Math.round(seconds * 10) / 10;
    } else {
      seconds = Math.round(seconds) / 10;
    }

    console.log("".concat(message, " (").concat(seconds, "s)"));
  }
}

function debounce(func, wait, immediate) {
  var timeout;
  return function () {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var later = function later() {
      timeout = null;
      if (!immediate) func.apply(void 0, args);
    };

    var callNow = immediate && !timeout;
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
    if (callNow) func.apply(void 0, args);
  };
}

var escapeReg = /[\\^$.*+?()[\]{}|]/g;

function escapeRegExp(str) {
  return str.replace(escapeReg, '\\$&');
}
